ee/src/main/java/org/jboss/as/ee/concurrent/ManagedThreadFactoryImpl.java

```diff
@@ -94,7 +111,33 @@ private CreateThreadAction(Runnable r, ContextHandle contextHandleForSetup) {
 
         @Override
         public AbstractManagedThread run() {
-            return ManagedThreadFactoryImpl.super.createThread(r, contextHandleForSetup);
+            return new ManagedThread(r, contextHandleForSetup);
+        }
+    }
+
+    /**
+     * Managed thread extension, to allow canceling the task running in the thread.
+     * @author emmartins
+     */
+    public class ManagedThread extends org.glassfish.enterprise.concurrent.ManagedThreadFactoryImpl.ManagedThread {
+        volatile ManagedFutureTask task = null;
+        /**
+         *
+         * @param target
+         * @param contextHandleForSetup
+         */
+        public ManagedThread(Runnable target, ContextHandle contextHandleForSetup) {
+            super(target, contextHandleForSetup);
+        }
+        /**
+         * Cancel the task running in the thread.
+         * @return
+         */
+        boolean cancelTask() {
+            if (task != null) {
+                return task.cancel(true);
+            }
```

Comment:
This is vulnerable to an NPE in a race.  Assigning 'task' a local var 'localTask' and then working with that eliminates that possibility.

Reply:
Fixed, and good catch, code was copied from the RI class it extends and didn't noticed. RI issue reported at https://github.com/eclipse-ee4j/concurrency-ri/issues/47

Reply:
👍